test/test.sh

```diff
@@ -0,0 +1,109 @@
+#! /bin/bash
+
+# wgetpaste test script
+# Exit code: number of mismatched downloads or 1 for general failure
+# Copyright (C) 2021  xxc3nsoredxx
+
+# Don't assume the test is being run from the same directory as the script
+TEST_DIR="$(dirname "$0")"
+TEST_FILE="$TEST_DIR/test.txt"
+DL_DIR="$(mktemp -q -d /tmp/wgetpaste_test.XXXXX)"
+# Services to hard skip
+# codepad: timing out
+HARD_SKIPS=('codepad')
+HARD_SKIP_COUNT=0
+AUTH_SKIP_COUNT=0
+FAIL_SKIP_COUNT=0
+DL_COUNT=0
+DL_MISMATCH=0
+
+# Test that temp directory was created
+if [ ! -d "$DL_DIR" ]; then
+    echo "Failed to create temporary download directory: $DL_DIR"
+    exit 1
+fi
+echo "Using download directory: $DL_DIR"
+
+# Post test file into each service (if possible)
+# Download the resulting paste into /tmp/wgetpaste_test.XXXXX/<service>.txt
+for serv in $("$TEST_DIR"/../wgetpaste -S --completions); do
+    # Hard skips
+    for hs in "${HARD_SKIPS[@]}"; do
+        if [ "$serv" == "$hs" ]; then
+            echo "HARD SKIP on $serv..."
+            HARD_SKIP_COUNT=$((HARD_SKIP_COUNT + 1))
+            continue 2
+        fi
+    done
+
+    # Log errors to analyze the reason
+    # Use verbose output to get more meaningful errors
+    # Log deleted at the end of each loop unless error other than 401
+    echo -n "Posting to $serv: "
+    ERROR_LOG="$DL_DIR/$serv-error.log"
+    URL="$("$TEST_DIR"/../wgetpaste -r -s "$serv" -v "$TEST_FILE" 2>"$ERROR_LOG")"
+    STATUS="$?"
+
+    # Skip failed posts (eg, not authorized for GitHub/GitLab, service error)
+    if [ "$STATUS" -ne 0 ]; then
+        if (grep -iq "HTTP.*401.*Unauthorized" "$ERROR_LOG"); then
+            echo "SKIPPING, needs authorization..."
```

Comment:
I think we should have a AUTH_SKIP_LIST similar to what we have for HARD_SKIP, otherwise services could silently break (e.g. if a service suddenly starte requiring auth) and the tests would still pass.

Reply:
Alright, I have it test each received 401 response against the services in `AUTH_SKIPS` to see if it's expected behavior or not.